File: 프로그래머스/unrated/258711. 도넛과 막대 그래프/도넛과 막대 그래프.py
# 도넛 모양 : 정점 개수 == 간선 개수
# 막대 모양 : 정점 개수 == 간선개수 + 1
# 8자 모양 : 정점 개수 == 간선개수 - 1
# 나가는 간선 개수가 2개 이상이거나 들어오는 간선이 없으면 생성한 정점
from collections import defaultdict, deque

graph = defaultdict(list)
def solution(edges):
    get_edge = defaultdict(bool)
    도넛, 막대, 팔자 = 0, 0, 0
    
    for a, b in edges:
        graph[a].append(b)
        get_edge[b] = True
    
    # 생성한 정점 찾기
    for node in graph.keys():
        if not get_edge[node] and len(graph[node]) >= 2:
            new_node = node
            break
            
        if not get_edge[node]:
            new_node = node
    
    for node in graph[new_node]:
        edge, node = count(node, 0, 1)
        if node == edge:
            도넛 += 1
        elif node == edge + 1:
            막대 += 1
        elif node == edge - 1:
            팔자 += 1

    return [new_node, 도넛, 막대, 팔자]

def count(now, edge, node):
    q = deque([now])
    visited_node = defaultdict(bool)
    visited_edge = defaultdict(bool)
    visited_node[now] = True
    
    while q:
        now = q.pop()
        for nxt in graph[now]:
            if not visited_node[nxt]:
                visited_node[nxt] = True
                node += 1
                q.append(nxt)
            if not visited_edge[(now, nxt)]:
                visited_edge[(now, nxt)] = True
                edge += 1
                q.append(nxt)
    return (edge, node)

File: 프로그래머스/2/258711. 도넛과 막대 그래프/도넛과 막대 그래프.java
// 나가는 것만 있고, 나가는 것의 개수가 가장 많은 것이 생성한 정점
// 도넛: 노드 개수 == 전선 개수
// 막대: 노드 개수 - 1  == 전선 개수
// 팔자: 노드 개수 + 1 == 전선개수

import java.util.*;

class Solution {
    int newNode = 0;
    int donut = 0;
    int rod = 0;
    int eight = 0;
    HashMap<Integer, ArrayList<Integer>> graph = new HashMap<>();
    
    public int[] solution(int[][] edges) {
        HashMap<Integer, Boolean> inMap = new HashMap<>();
        
        for (int[] edge: edges) {
            int start = edge[0];
            int end = edge[1];
            ArrayList<Integer> temp = graph.getOrDefault(start, new ArrayList<>());
            temp.add(end);
            graph.put(start, temp);
            inMap.put(end, true);
        }
        
        int inMaxCount = 0;
        for (int i: graph.keySet()) {
            if (inMap.containsKey(i)) continue;
            if (graph.get(i).size() > inMaxCount) {
                inMaxCount = graph.get(i).size();
                newNode = i;
            }   
        }
        for (int i: graph.get(newNode)) {
            Count count = getCount(i);
            if (count.edge == count.node) {
                donut++;
            } else if (count.edge == count.node - 1) {
                rod++;
            } else if (count.edge == count.node + 1) {
                eight++;
            }
        }
        
        int[] answer = {newNode, donut, rod, eight};
        return answer;
    }
    
    Count getCount(int start) {
        int node = 1;
        int edge = 0;
        
        ArrayDeque<Integer> dq = new ArrayDeque<>();
        HashMap<Integer, Boolean> visited = new HashMap<>();
        dq.add(start);
        visited.put(start, true);
        while (!dq.isEmpty()) {
            int now = dq.removeFirst();
            for (int next: graph.getOrDefault(now, new ArrayList<Integer>())) {
                edge++;
                if (!visited.containsKey(next)) {
                    visited.put(next, true);
                    node++;
                    dq.addLast(next);
                }
            }
        }
        return new Count(node, edge);
    }
    
    class Count {
        int node;
        int edge;
        
        Count(int node, int edge) {
            this.node = node;
            this.edge = edge;
        }
    }
}